### PR TITLE
Fix ape escape crash, optimize IR memory usage

### PR DESCRIPTION
**Pull request by ZachCook**

### CODE REVIEW

**Comment by pcercuei** [comment]
> When threaded compiler is enabled the smaller flags size does actually save 8 bytes on 64 bit, since the non-pointer sizes are as follows with brackets showing 8 byte alignment: (4,4),(4,2,1,1) , otherwise it ends up as (4,4),(4,2,2),(1,7padding)

AFAIK the size of `atomic_flag` is not defined in the standard, so this only works if it's one byte. Thanksfully this seems to be the case at least on Linux.

**Comment by ZachCook** [comment]
I also have a patch I'm experimenting with that moves the struct opcode's flags into a separate array in the block, which roughly cuts struct opcode memory usage by 25% or even 50% if I get rid of LIGHTREC_SMC flag (relying on the block's BLOCK_NEVER_COMPILE flag instead, so invalidating on any io in blocks containing SMC instead of just the self modifying io, not sure how much worse that is or if it is fully correct to do so?), but it makes things quite a bit more complicated, especially if wanting avoid adding a separate pointer by tacking it on to the end of the opcode_list.

I see about very roughly half of IR memory being struct block, and the other half is struct opcode, so either 12.5% or 25% reduction in IR memory if sharing the opcode pointer, and if adding a seperate pointer only removing SMC flag actually saves 12.5% IR memory.

Is that worth the possible memory savings?

**Comment by pcercuei** [comment]
There will eventually be other flags, so I don't think you should limit the flags field to a byte. That simplifies the LIGHTREC_SMC situation. You could still save the 25% by using a u16 array in the block, yes. Whether or not that's a good idea, I wouldn't know... That would save a bit of RAM but maybe the cache locality would be worse. Feel free to try!

**Comment by pcercuei** [comment]
About the third commit: it would probably also be a good idea to have a helper function to get a memory map + host address from a PSX address, since the same code is used at many different places (lightrec_memset, lightrec_rw, lightrec_read_opcode, lightrec_precompile_block, lightrec_invalidate).

Btw, the block LUT contains pointers, so the size of the block is irrelevant, and the LUT size didn't change with your update.

Could you sign your commits (git commit --amend -s) so that I can merge them?

Thanks!

File: lightrec.c [hunk 1]
```diff
@@ -199,44 +199,48 @@ static void lightrec_invalidate_map(struct lightrec_state *state,
...
 	}
 
-	return NULL;
+	if(i == state->nb_maps)
+		return NULL;
+
+	while (map->mirror_of)
+		map = map->mirror_of;
+
+	*host = map->address + kaddr - map->pc;
```
Comment (pcercuei) [inline lightrec.c]: You could check `if (host)` first, then callers could pass NULL if they are not interested in the host address.

File: lightrec-private.h [hunk 2]
```diff
@@ -49,6 +49,9 @@
 #define BLOCK_IS_DEAD		BIT(3)
 #define BLOCK_IS_MEMSET		BIT(4)
 
+/* USE A LARGER TYPE IF A BIT(8) FLAG IS ADDED IN FUTURE */
+#define block_flags_t u8
```
Comment (pcercuei) [inline lightrec-private.h]: typedef?

**Review by pcercuei — CHANGES_REQUESTED** [review]
Need a Signed-off-by line.

**Comment by pcercuei** [comment]
Note for later - you could totally have the opcode list (which is more an array than a list now) at a `u32 opcodes[];` at the end of `struct block`, and allocate the block accordingly. That would save a pointer from `struct block;`, cause less malloc overhead, and there wouldn't be any pointer indirection needed to read an opcode from the `struct block` pointer.

File: lightrec-private.h (hunk 2 above)
Comment (ZachCook) [reply lightrec-private.h]: actually could just get rid of it instead as nothing else uses it other than the struct definition

Comment (pcercuei) [reply lightrec-private.h]: Yep. And the comment can go away, if you add a BIT(8) later, the compiler will complain and we can update to a u16 then.

**Review by pcercuei — APPROVED** [review]

**Comment by pcercuei** [comment]
Awesome work, thanks!

**Pull request merged**